### ISSUE TITLE
`needless_pass_by_ref_mut` false positive in nested async

### ISSUE THREAD

**Issue by hatoo** [rust-lang/rust-clippy#11299]
### Summary

`needless_pass_by_ref_mut` false positive

I think it's another variant of #11179, sorry if this issue is a duplicate.

### Lint Name

needless_pass_by_ref_mut

### Reproducer

I tried this code:

```rust
async fn f(x: &mut i32) {
    async {
        *x += 1;
    }
    .await;
}
```

I saw this happen:

```
❯ cargo +nightly clippy 
warning: this argument is a mutable reference, but not used mutably
 --> src/main.rs:8:15
  |
8 | async fn f(x: &mut i32) {
  |               ^^^^^^^^ help: consider changing to: `&i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
  = note: `#[warn(clippy::needless_pass_by_ref_mut)]` on by default
```

I expected to see this happen:

No warning

### Version

```text
rustc 1.73.0-nightly (eb088b8b9 2023-08-05)
binary: rustc
commit-hash: eb088b8b9d98f1af1b0e61bbdcd8686e1b0db7b6
commit-date: 2023-08-05
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```

**Comment by Centri3** [comment]
cc @GuillaumeGomez

Doesn't look like the same issue.

**Comment by GuillaumeGomez** [comment]
I see the problem: need recursion for closures. Gonna check tomorrow.

**Comment by GuillaumeGomez** [comment]
I opened https://github.com/rust-lang/rust-clippy/pull/11314 to fix this issue.

**Comment by MingweiSamuel** [comment]
Following example is also a false positive:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2f930223459b2cb29f6c9e4a06a87546
```rust
pub async fn foo(n: &mut usize) -> impl '_ + FnMut() {
    || {
        *n += 1;
    }
}
```
@GuillaumeGomez could you add [something like] this to the test cases in #11314 ?

**Comment by SpeckyYT** [comment]
I'm also experiencing this same false positive.
```
warning: this argument is a mutable reference, but not used mutably
 --> src\input.rs:8:19
  |
8 |     inputs_queue: &mut VecDeque<Inputs<bool>>,
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider changing to: `&VecDeque<Inputs<bool>>`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
  = note: `#[warn(clippy::needless_pass_by_ref_mut)]` on by default
  ```
  despite doing `inputs_queue.append(&mut inputs_vector)` in code.
  I also use async/await, probably that's what is causing this false positive?

**Comment by GuillaumeGomez** [comment]
@MingweiSamuel Thanks! I'll add it.

@SpeckyYT Would have been better to come up with a small example showing the issue. ^^'

**Comment by SpeckyYT** [comment]
Sorry for not providing an example.
Here is a simplified version of what my code is doing (the original code contains `.await`s and other stuff, but they seem to not affect this bug).
```rs
async fn a(b: &mut Vec<bool>) {
    b.append(&mut vec![])
}
```
```
warning: this argument is a mutable reference, but not used mutably
 --> src\test.rs:1:15
  |
1 | async fn a(b: &mut Vec<bool>) {
  |               ^^^^^^^^^^^^^^ help: consider changing to: `&Vec<bool>`
  |
  ```

**Comment by GuillaumeGomez** [comment]
Thanks! I don't know if #11314 is the one who fixed it but in my case I don't have any warning with your code. I tested with `push` as well just in case. I'll add your code as a regression test in the PR.

**Comment by MingweiSamuel** [comment]
> @MingweiSamuel Just took a look and your case is a different one. Do you mind opening a new issue and tagging me on it please?

Looks like someone opening https://github.com/rust-lang/rust-clippy/issues/11380 for this
